packages/nextjs/hooks/scaffold-eth/useScaffoldEventHistory.ts

```diff
@@ -45,135 +80,82 @@ export const useScaffoldEventHistory = <
-                transactionData && logs[i].transactionHash !== null
-                  ? await publicClient.getTransaction({ hash: logs[i].transactionHash as Hash })
-                  : null,
-              receipt:
-                receiptData && logs[i].transactionHash !== null
-                  ? await publicClient.getTransactionReceipt({ hash: logs[i].transactionHash as Hash })
-                  : null,
-            });
-          }
-          setEvents([...newEvents, ...events]);
-          setError(undefined);
-        }
-      } catch (e: any) {
-        if (events.length > 0) {
-          setEvents([]);
-        }
-        setError(e);
-        console.error(e);
-      } finally {
-        setIsLoading(false);
-      }
-    },
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [
-      blockData,
-      deployedContractData,
-      enabled,
-      eventName,
-      events,
-      // eslint-disable-next-line react-hooks/exhaustive-deps
-      JSON.stringify(filters, replacer),
-      fromBlockUpdated,
-      publicClient,
-      receiptData,
-      transactionData,
+  const [isFirstRender, setIsFirstRender] = useState(true);
+
+  const { data: blockNumber } = useBlockNumber({ watch: watch, chainId: targetNetwork.id });
+
+  const { data: deployedContractData } = useDeployedContractInfo(contractName);
+
+  const event =
+    deployedContractData &&
+    ((deployedContractData.abi as Abi).find(part => part.type === "event" && part.name === eventName) as AbiEvent);
+
+  const query = useInfiniteQuery({
+    queryKey: [
+      "eventHistory",
+      {
+        contractName,
+        address: deployedContractData?.address,
+        eventName,
+        fromBlock: fromBlock.toString(),
+        chainId: targetNetwork.id,
+      },
     ],
-  );
-
-  useEffect(() => {
-    if (!deployedContractLoading) {
-      readEvents();
-    }
-  }, [readEvents, deployedContractLoading]);
-
-  useEffect(() => {
-    // Reset the internal state when target network or fromBlock changed
-    setEvents([]);
-    setFromBlockUpdated(fromBlock);
-    setError(undefined);
-  }, [fromBlock, targetNetwork.id]);
-
-  useInterval(
-    async () => {
-      if (!deployedContractLoading) {
-        readEvents();
-      }
+    queryFn: async ({ pageParam }) => {
+      if (!Boolean(deployedContractData?.address) || !Boolean(publicClient)) return undefined;
+      const data = await getEvents(
+        { address: deployedContractData?.address, event, fromBlock: pageParam, args: filters },
+        publicClient,
+        { blockData, transactionData, receiptData },
+      );
+
+      return data;
     },
-    watch && enabled ? (targetNetwork.id !== chains.hardhat.id ? scaffoldConfig.pollingInterval : 4_000) : null,
-  );
-
-  const eventHistoryData = useMemo(
-    () =>
-      events?.map(addIndexedArgsToEvent) as UseScaffoldEventHistoryData<
+    enabled: enabled && Boolean(deployedContractData?.address) && Boolean(publicClient),
```

Comment:
nit: Boolean(deployedContractData?.address) && Boolean(publicClient) can be moved to variable and also used as negative here https://github.com/scaffold-eth/scaffold-eth-2/pull/869/files#diff-0efe8962fd8674c8a04310503b57e5e0ff36eb7016ae11764e1cc83472bc7646R109

Reply:
Ohh nice yup, updated at https://github.com/scaffold-eth/scaffold-eth-2/pull/869/commits/30ed018465f54b5f2a7de5ec26c1f25bb5a131df 🙌